### PR TITLE
feat: try to support custom OAuth provider

### LINKED ISSUE #655: use custom OAuth provider
I want to use the private OAuth server by setting provider in casdoor. And I'm not sure this is a general requirement.

### PR DESCRIPTION
**Pull request by halozhy**
try to fix #655
I create an oauth provider type called Private, which requires 5 extra params: `AuthPage`, `Scope`, `TokenAPI`, `UserInfo API` and `Icon` 
`AuthPage` is the login page for the private provider
`Scope` is a carrying parameter when accessing `AuthPage`
`Token API`  is a private provider's API that exchanges code for accessToken
`UserInfo API` is a private provider's API that uses accesstoken to obtain user information
`Icon` is an icon link that defines the private provider's icon

I assume Google, GitHub and Casdoor are private providers respectively and fill in their parameters. After testing, it can log in successfully

For more detailed info, please see https://gist.github.com/halozhy/345944711e4bfd1f51e1ff7cc4570f12 (I wrote it in Chinese)

### CODE REVIEW

**Comment by hsluoyz** [comment]
@seriouszyx plz review

File: idp/private.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+func NewPrivateIdProvider(clientId string, clientSecret string, redirectUrl string, authPage string, tokenApi string, userInfoApi string) *PrivateIdProvider {
+	idp := &PrivateIdProvider{}
+	idp.UserInfoApi = userInfoApi
+
+	var config = &oauth2.Config{
+		ClientID:     clientId,
+		ClientSecret: clientSecret,
+		RedirectURL:  redirectUrl,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  authPage,
```
Comment (hsluoyz) [inline idp/private.go:44]: AuthUrl

File: idp/private.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+			TokenURL: tokenApi,
```
Comment (hsluoyz) [inline idp/private.go:45]: tokenUrl

File: controllers/auth.go [hunk 3]
```diff
@@ -151,7 +151,7 @@ func (c *ApiController) GetApplicationLogin() {
 }
 
 func setHttpClient(idProvider idp.IdProvider, providerType string) {
-	if providerType == "GitHub" || providerType == "Google" || providerType == "Facebook" || providerType == "LinkedIn" || providerType == "Steam" {
+	if providerType == "GitHub" || providerType == "Google" || providerType == "Facebook" || providerType == "LinkedIn" || providerType == "Steam" || providerType == "Private" {
```
Comment (hsluoyz) [inline controllers/auth.go:154]: Don't add this

File: idp/private.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	idp.Client = client
+}
+
+func (idp *PrivateIdProvider) GetToken(code string) (*oauth2.Token, error) {
+	ctx := context.WithValue(context.Background(), oauth2.HTTPClient, idp.Client)
+	return idp.Config.Exchange(ctx, code)
+}
+
+func (idp *PrivateIdProvider) GetUserInfo(token *oauth2.Token) (*UserInfo, error) {
+	cdUserinfo := &CasdoorUserInfo{}
```
Comment (hsluoyz) [inline idp/private.go:63]: It doesn't have to be `CasdoorUserInfo`

File: idp/provider.go [hunk 5]
```diff
@@ -35,7 +35,7 @@ type IdProvider interface {
 	GetUserInfo(token *oauth2.Token) (*UserInfo, error)
 }
 
-func GetIdProvider(typ string, subType string, clientId string, clientSecret string, appId string, redirectUrl string, hostUrl string) IdProvider {
+func GetIdProvider(typ string, subType string, clientId string, clientSecret string, appId string, redirectUrl string, hostUrl string, authPage string, tokenApi string, userInfoApi string) IdProvider {
```
Comment (hsluoyz) [inline idp/provider.go:38]: userInfoUrl

File: web/src/Setting.js [hunk 6]
```diff
@@ -414,6 +414,7 @@ export function getProviderTypeOptions(category) {
         {id: 'AzureAD', name: 'AzureAD'},
         {id: 'Slack', name: 'Slack'},
         {id: 'Steam', name: 'Steam'},
+        {id: 'Private', name: 'Private'},
```
Comment (hsluoyz) [inline web/src/Setting.js:417]: Private -> Custom

File: object/provider.go [hunk 7]
```diff
@@ -27,16 +27,21 @@ type Provider struct {
...
+	Category           string `xorm:"varchar(100)" json:"category"`
+	Type               string `xorm:"varchar(100)" json:"type"`
+	SubType            string `xorm:"varchar(100)" json:"subType"`
+	Method             string `xorm:"varchar(100)" json:"method"`
+	ClientId           string `xorm:"varchar(100)" json:"clientId"`
+	ClientSecret       string `xorm:"varchar(2000)" json:"clientSecret"`
+	ClientId2          string `xorm:"varchar(100)" json:"clientId2"`
+	ClientSecret2      string `xorm:"varchar(100)" json:"clientSecret2"`
+	Cert               string `xorm:"varchar(100)" json:"cert"`
+	PrivateAuthPage    string `xorm:"varchar(200)" json:"privateAuthPage"`
```
Comment (hsluoyz) [inline object/provider.go:40]: All `Private` -> `Custom`

File: object/provider.go [hunk 8]
```diff
@@ -27,16 +27,21 @@ type Provider struct {
...
+	PrivateScope       string `xorm:"varchar(200)" json:"privateScope"`
+	PrivateTokenApi    string `xorm:"varchar(200)" json:"privateTokenApi"`
+	PrivateUserInfoApi string `xorm:"varchar(200)" json:"privateUserInfoApi"`
+	PrivateIcon        string `xorm:"varchar(200)" json:"privateIcon"`
```
Comment (hsluoyz) [inline object/provider.go:44]: CustomLogo

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by seriouszyx** [comment]
@halozhy Hi, I pull the codes and create a private provider as below.

image

But I got an unclear error message after I finished authentication in the provider. 

image

**Comment by halozhy** [comment]
> But I got an unclear error message after I finished authentication in the provider.

@seriouszyx Sorry, I made a typo in my gist screenshot. The Token API `http://1.117.101.146:7001/login/oauth/access_token` should be changed to `http://1.117.101.146:7001/api/login/oauth/access_token`

File: web/src/ProviderEditPage.js [hunk 9]
```diff
@@ -256,6 +256,79 @@ class ProviderEditPage extends React.Component {
             </React.Fragment>
           )
         }
+        {
+          this.state.provider.type !== "Custom" ? null : (
+            <React.Fragment>
+              <Row style={{marginTop: '20px'}} >
+                <Col style={{marginTop: '5px'}} span={(Setting.isMobile()) ? 22 : 2}>
+                  {Setting.getLabel(i18next.t("provider:Auth URL"), i18next.t("provider:Auth URL - Tooltip"))}
```
Comment (hsluoyz) [inline web/src/ProviderEditPage.js:264]: Can you provide example values for these CustomXXX params? You can directly use: https://door.casdoor.com/ values. https://door.casdoor.com/ will support to be connected. We use real examples.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/ProviderEditPage.js (hunk 9 above)
Comment (halozhy) [reply web/src/ProviderEditPage.js:270]: In commit c021fd834c6ccaf620e575a77d08cd6b509bb11e, I set some example values for customAuthUrl, customScope, customTokenUrl and customUserInfoUrl

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**